APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+surround the matrix with the ``#fmt: off``/``#fmt: on`` escape hatch, as this
+clearly makes the code more readable.  While many examples are less clear than
+the above, and indeed in most cases "readability" is a subjective concept, this
+mechanism allows for such cases to be explicitly excepted where the developer
+thinks the improved readability warrants it.
+
+History
+~~~~~~~
+This changes almost every Python file. Git blame can skip the commit, but git
+diff won't. But we are already doing other bulk changes that hit every file.
+
+Code correctness
+~~~~~~~~~~~~~~~~
+Black guarantees correctness of the code formatter to not change the code AST.
+
+Migration effort
+~~~~~~~~~~~~~~~~
+Enthusiastic champions of the effort will largely do the work, but we recognize
+and appreciate that additional effort will be required from maintainers that are
+neutral or against the adoption of Black.
+
+Backports / release branches
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This is addressed in the Implementation section.
+
+New contributors
+~~~~~~~~~~~~~~~~
+New contributors may be more discouraged by more steps. However:
+
+This *replaces* the existing PEP8 check, so that it is not "another step" To
+make it better, leave very clear explicit messages in the status checks on GH
+(not putting up a barrier) Scikit-learn had a few issues with contributors but
+not at a level that caused regret about the decision to adopt black formatting.
+One developer described the change as a "net benefit".
+
+Removing the humanity
+~~~~~~~~~~~~~~~~~~~~~
+Some developers may feel that individual style and humanity gets lost in
+autoformatting. We hope to convey that humanity and individuality is expressed
+more fully in the elegance and clarity of the code logic and implementation. In
+most cases the placement of white space and line breaks can be viewed as a
+mechanistic process that humans need not control.
+
+Exception Policy
+^^^^^^^^^^^^^^^^
+As described above, there are cases Black formatting makes the code
+substantially more difficult to read or understand.. These should be treated
+explicitly as exceptions, using the ``#fmt: off``/``#fmt: on`` escape hatch for
+multi-line exceptions, or ``#fmt: skip`` for single-line exceptions. **This is
+allowed if the code author and PR reviewer(s) agree that the code line or block
+should be excluded from Black formatting.**
+
+
+Branches and pull requests
+--------------------------
+
+- `Add configuration for black autoformatter <https://github.com/astropy/astropy/pull/13253>`_
+- `Apply black to modeling <https://github.com/astropy/astropy/pull/13254>`_
+
+Implementation
+--------------
+
+All Python code in ``astropy`` is formatted with Black_, using its default
+settings, that is, 88 characters per line and double quotes. Note that, by
+design, most of Black_ is inflexible, and there is not a way to pick-and-choose
+some elements and not others of Black's format.
+
+Implementing this change requires:
```

Comment:
PR astropy/astropy#13253 sets up the changes to `astropy` to enable enforcing `Black` on a per-subpackage basis, but does not run black on any part of `astropy`. Note that the changes to look at are to the `.pre-commit-config.yaml` and `pyproject.toml` files, all the other file changes are due to a necessary change to the configuration for `isort` so that it is compatible with `Black`.

One minor change to this PR, so that it fits the currently proposed APE, is updating the `line_length` values (for `flake8`, `isort`, and `Black`) so that they all conform to the default `Black` line length of 88.

Reply:
Or changing Black to a line length of 100 (https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html?highlight=length#line-length)

Reply:
Default (88) please! :smile:  I've been using Black with 88 and I like it.

Reply:
I heard we were painting bike sheds. Longer lines ftw.

Reply:
Split the difference, and go with 94?

Reply:
For my exact laptop and editor configuration, I can get two screens of 84 characters split vertically. But to be honest I very rarely do that split, so I'm fine with any value between 88 and 94. I ran black with a width of 94 and that does reduce the diffs a bit.

Reply:
I'm used to the default 88 but I don't feel very strongly.

(When I first read this APE, I was still pro-`yapf` because of the customizable features (esp. around array reformatting), but now that I see how much discussion we would have around each setting, I'm glad `black` only has a few 😅.)

Reply:
88 is plenty, I see no reason not to use the default value.

Reply:
I don't really mind either way, my suggestion of 94 above wasn't necessarily very serious.

Reply:
OK, I'll revert back to 88. I don't mind, I just want *something*. :smile: But personally I like the simplicity of using the default.